Allow predicates to disabled on a case-by-case basis via annotation. (#5284)
elastic/cloud-on-k8s: docs/orchestrating-elastic-stack-applications/elasticsearch/orchestration.asciidoc

** Adjust the Elasticsearch link:https://www.elastic.co/guide/en/elasticsearch/reference/current/indices-update-settings.html[index settings] to a number of replicas that allow the desired node removal.
** Use link:https://www.elastic.co/guide/en/elasticsearch/reference/current/index-modules.html#dynamic-index-settings[`auto_expand_replicas`] to automatically adjust the replicas to the number of data nodes in the cluster.

[id="{p}-advanced-upgrade-control"]
== Advanced control during rolling upgrades

The rules (otherwise known as predicates) that the ECK operator follows during an Elasticsearch upgrade can be selectively disabled for certain scenarios where the ECK operator will not proceed with an Elasticsearch cluster upgrade because it deems it to be "unsafe".

WARNING: Selectively disabling the predicates below are extremely risky, and carry a high chance of either data loss, or causing a cluster to become completely unavailable.  Use only if you are absolutely positive you are not causing permanent damage to an Elasticsearch cluster.

> * The following named predicates control the upgrade process

Reviewer: Maybe also mention that all the predicates can be disabled using `"*"`

Author: good point.

** do_not_restart_healthy_node_if_MaxUnavailable_reached
+
If `maxUnavailable` is reached, only allow unhealthy Pods to be deleted.
** skip_already_terminating_pods
+
Do not attempt to restart pods that are already in the process of being terminated.
** only_restart_healthy_node_if_green_or_yellow
+
Only restart healthy Elasticsearch nodes if the health of the cluster is either green or yellow, never red.
** if_yellow_only_restart_upgrading_nodes_with_unassigned_replicas
+
During a rolling upgrade, primary shards assigned to a node running a new version cannot have their replicas assigned to a node with the old version. Therefore we must allow some Pods to be restarted even if cluster health is Yellow so the replicas can be assigned.
** require_started_replica
+
If a cluster is yellow, allow deleting a node, but only if they do not contain the only replica of a shard since it would make the cluster go red.
** one_master_at_a_time
+
Only allow a single master to be upgraded at a time.
** do_not_delete_last_master_if_all_master_ineligible_nodes_are_not_upgraded
+
Force an upgrade of all the master-ineligible nodes before upgrading the last master-eligible node.

> ** do_not_delete_pods_with_same_shards

Reviewer 2: As the predicate names become public with this feature, I wonder if we should write a little check to run in CI to verify that the names in the doc are in sync with the names in the code.

Something like this:
```bash
code=pkg/controller/elasticsearch/driver/upgrade_predicates.go \
doc=docs/orchestrating-elastic-stack-applications/elasticsearch/orchestration.asciidoc \
predicate_pattern='[a-z]*_[A-Za-z_]*'
diff \
	<(grep "name:" "$code" | grep -o "$predicate_pattern" ) \
	<(grep '\*\* [a-z]' "$doc" | grep -o "$predicate_pattern" )
```
If we add a comment in the two files to mark the predicate name (`// predicate`), this could make the parsing more sure. 
```
diff \
	<(grep '// predicate' $code | grep -o "$predicate_pattern" ) \
	<(grep '// predicate' -A1 $doc | grep -o "$predicate_pattern" )
```

Author: This is a good thought. I've added this CI check, missing the marks.  Let's see if that's sufficient.  I'm happy to also add the marks if we feel that's necessary.

WARNING: Selectively disabling the predicates below are extremely risky, and carry a high chance of either data loss, or causing a cluster to become completely unavailable. Use only if you are absolutely positive you are not causing permanent damage to an Elasticsearch cluster. Note that these predicates have the potential to change in the future. We will be adding, removing, and renaming these over time, so take great care in adding these to any automation.  Also please ensure you remove them after use. `kublectl annotate elasticsearch.elasticsearch.k8s.elastic.co/elasticsearch-sample eck.k8s.elastic.co/disable-upgrade-predicates-`

* The following named predicates control the upgrade process

** do_not_restart_healthy_node_if_MaxUnavailable_reached
+
If `maxUnavailable` is reached, only allow unhealthy Pods to be deleted.
** skip_already_terminating_pods
+
Do not attempt to restart pods that are already in the process of being terminated.
** only_restart_healthy_node_if_green_or_yellow
+
Only restart healthy Elasticsearch nodes if the health of the cluster is either green or yellow, never red.
** if_yellow_only_restart_upgrading_nodes_with_unassigned_replicas
+

> During a rolling upgrade, primary shards assigned to a node running a new version cannot have their replicas assigned to a node with the old version. Therefore we must allow some Pods to be restarted even if cluster health is Yellow so the replicas can be assigned.

Reviewer 3: ```suggestion
During a rolling upgrade, primary shards assigned to a node running a new version cannot have their replicas assigned to a node with the old version. Therefore we must allow some Pods to be restarted even if cluster health is yellow so the replicas can be assigned.
```

** require_started_replica
+
If a cluster is yellow, allow deleting a node, but only if they do not contain the only replica of a shard since it would make the cluster go red.
** one_master_at_a_time
+
Only allow a single master to be upgraded at a time.
** do_not_delete_last_master_if_all_master_ineligible_nodes_are_not_upgraded
+
Force an upgrade of all the master-ineligible nodes before upgrading the last master-eligible node.
** do_not_delete_pods_with_same_shards
+

> Do not allow 2 pods to be deleted at the same time, that contain the same shard.

Reviewer 3: ```suggestion
Do not allow two pods containing the same shard to be deleted at the same time.
```

Any of these predicates can be disabled by adding an annotation with the key of `eck.k8s.elastic.co/disable-upgrade-predicates` to the Elasticsearch metadata, specifically naming the predicate that is needing to be disabled.  Also, all predicates can be disabled by replacing the name of any predicatae with "*".

* Example use case

> Assume a given Elasticsearch cluster is a "red" state because of ab un-allocatable shard setting that was applied to the cluster

Reviewer 3: ```suggestion
Assume a given Elasticsearch cluster is a "red" state because of an un-allocatable shard setting that was applied to the cluster:
```

```json
{
	"settings": {
		"index.routing.allocation.include._id": "does not exist"
	}
}
```

This cluster would never be allowed to be upgraded with the standard set of upgrade predicates in place, as the cluster is in a "red" state, and the named predicate `only_restart_healthy_node_if_green_or_yellow` prevents the upgrade.

If the following annotation was added to the cluster specification, and the version was increased from 7.15.2 -> 7.15.3

```yaml
apiVersion: elasticsearch.k8s.elastic.co/v1
kind: Elasticsearch
metadata:
  name: testing
  annotations:
    eck.k8s.elastic.co/disable-upgrade-predicates: "only_restart_healthy_node_if_green_or_yellow"
    # Also note that eck.k8s.elastic.co/disable-upgrade-predicates: "*" would work as well, but is much less selective.
spec:
  version: 7.15.3 # previously set to 7.15.2, for example
```

> The ECK operator would allow this upgrade to proceed, even though the cluster was in a "red" state during this upgrade process.

Reviewer: nit: maybe add another warning: "Do not forget to remove the annotation once not needed" , and a one-liner to remove it: `kublectl annotate elasticsearch.elasticsearch.k8s.elastic.co/elasticsearch-sample eck.k8s.elastic.co/disable-upgrade-predicate-`
